File: py/src/modular_api/core/module_builder.py
"""Fluent builder that registers use cases on a Starlette Router.

Returned by ``ModularApi.module()`` inside the callback:

    api.module("users", lambda m: (
        m.usecase("create", CreateUser.from_json),
        m.usecase("list", ListUsers.from_json, method="GET"),
    ))

Each ``usecase()`` call mounts an endpoint on the module's Router
and stores metadata in ``api_registry`` for OpenAPI generation.

Mirror of ``ModuleBuilder`` in Dart and TypeScript.
"""

from __future__ import annotations

from typing import Any, Callable

from starlette.routing import Route, Router

from modular_api.core.registry import (
    UseCaseDocMeta,
    UseCaseFactory,
    UseCaseRegistration,
    api_registry,
)
from modular_api.core.usecase import Input, Output, UseCase
from modular_api.core.usecase_handler import usecase_handler


class ModuleBuilder:
    """Collects use cases for one module and mounts them on a Starlette Router.

    The builder provides a fluent API — every ``usecase()`` returns ``self``
    so multiple registrations can be chained.
    """

    def __init__(self, base_path: str, module_name: str) -> None:
        self._base_path = base_path
        self._module_name = module_name
        self._routes: list[Route] = []

    # ── Public API ────────────────────────────────────────────

    def usecase(
        self,
        name: str,
        factory: UseCaseFactory,
        *,
        method: str = "POST",
        summary: str | None = None,
        description: str | None = None,
    ) -> ModuleBuilder:
        """Register a use case as an HTTP endpoint on this module.

        Returns ``self`` for fluent chaining.
        """
        clean_name = name.strip().lstrip("/")
        method_upper = method.upper()
        full_path = f"{self._normalize_base(self._base_path)}/{self._module_name}/{clean_name}"

        # Mount endpoint on internal route list
        self._routes.append(
            Route(f"/{clean_name}", endpoint=usecase_handler(factory), methods=[method_upper]),
        )

        # Capture schemas via a dummy factory call (fail-safe)
        schemas = self._extract_schemas(factory)

        # Register metadata for OpenAPI generation
        doc = UseCaseDocMeta(
            summary=summary or f"Use case {clean_name} in module {self._module_name}",
            description=description or f"Auto-generated documentation for {clean_name}",
            tags=[self._module_name],
        )

        api_registry.routes.append(
            UseCaseRegistration(
                module=self._module_name,
                name=clean_name,
                method=method_upper,
                path=full_path,
                factory=factory,
                schemas=schemas,
                doc=doc,
            ),
        )

        return self

    def build_router(self) -> Router:
        """Create a Starlette Router containing all registered use-case routes."""
        return Router(routes=list(self._routes))

    # ── Private helpers ───────────────────────────────────────

    @staticmethod
    def _extract_schemas(factory: UseCaseFactory) -> dict[str, dict[str, Any]]:
        """Capture Input and Output schemas independently from a dummy factory call.

        Each DTO extraction has its own try/except so a failure in one
        (e.g. Output not yet initialised) does not destroy the other.
        This matches Dart's separate _inferInputSchema/_inferOutputSchema pattern.
        """
        input_schema: dict[str, Any] = {}
        output_schema: dict[str, Any] = {}

        try:
            instance = factory({})
        except Exception:
            # Factory itself failed — both schemas fall back to empty.
            return {"input": input_schema, "output": output_schema}

        try:
            input_schema = instance.input.to_schema()
        except Exception:
            # Input DTO inaccessible or to_schema() failed — keep empty fallback.
            pass

        try:
            output_schema = instance.output.to_schema()
        except Exception:
            # Output not initialised until execute() — expected for most UseCases.
            pass

        return {"input": input_schema, "output": output_schema}

    @staticmethod
    def _normalize_base(path: str) -> str:
        """Ensure base_path starts with '/' (or is empty)."""
        if not path:
            return ""
        return path if path.startswith("/") else f"/{path}"

File: ts/src/core/module_builder.ts
// ============================================================
// core/module_builder.ts
// ModuleBuilder — collects use cases and mounts them on a Router.
// Mirror of ModuleBuilder in Dart.
// ============================================================

import { Router } from 'express';
import type { Input, Output, UseCaseFactory } from './usecase';
import { UseCase } from './usecase';
import { useCaseHandler } from './usecase_handler';
import { apiRegistry } from './registry';
import { getFieldMetadata } from './schema/field';

type HttpMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';

export interface UseCaseOptions {
  /** HTTP method. Defaults to POST (same as Dart version). */
  method?: HttpMethod;
  summary?: string;
  description?: string;
  /** Override input schema for OpenAPI (if fromJson fails with empty data) */
  inputSchema?: Record<string, unknown>;
  /** Override output schema for OpenAPI (if fromJson fails with empty data) */
  outputSchema?: Record<string, unknown>;
  /** Input class for pre-validation and schema extraction (enables strict fromJson). */
  inputClass?: abstract new (...args: unknown[]) => Input;
  /** Output class for schema extraction. */
  outputClass?: abstract new (...args: unknown[]) => Output;
}

/**
 * Fluent builder that registers use cases on a module-scoped Express Router.
 * Returned and used inside the callback of `ModularApi.module()`.
 *
 * Dart equivalent:
 *   api.module('users', (m) {
 *     m.usecase('create', CreateUser.fromJson);
 *   });
 *
 * TypeScript:
 *   api.module('users', (m) => {
 *     m.usecase('create', CreateUser.fromJson);
 *   });
 */
export class ModuleBuilder {
  private readonly router: Router;

  constructor(
    private readonly basePath: string,
    private readonly moduleName: string,
    private readonly rootRouter: Router,
  ) {
    this.router = Router();
  }

  /**
   * Registers a use case as an HTTP endpoint.
   *
   * @param name     Route segment, e.g. 'create' → POST /api/users/create
   * @param factory  The static `fromJson` of your UseCase class
   * @param options  Optional HTTP method, summary and description for OpenAPI
   */
  usecase<I extends Input, O extends Output>(
    name: string,
    factory: UseCaseFactory<I, O>,
    options: UseCaseOptions = {},
  ): this {
    const { method = 'POST', summary, description, inputSchema, outputSchema, inputClass, outputClass } = options;

    // Normalize name: trim and remove leading slash
    const cleanName = name.trim().replace(/^\//g, '');
    const subPath = `/${cleanName}`;
    const methodL = method.toLowerCase() as Lowercase<HttpMethod>;

    // Mount the Express handler (with optional pre-validation)
    this.router[methodL](subPath, useCaseHandler(factory, { inputClass }));

    // Try to capture schemas — prefer class-level metadata, then dummy factory
    const extracted = this._extractSchemas(factory, inputClass, outputClass);
    const schemas = {
      input: inputSchema ?? extracted.input,
      output: outputSchema ?? extracted.output,
    };

    // Register in the global registry for OpenAPI generation
    apiRegistry.routes.push({
      module: this.moduleName,
      name: cleanName,
      method: method,
      path: `${this._normalizeBase(this.basePath)}/${this.moduleName}/${cleanName}`,
      factory: factory as UseCaseFactory<Input, Output>,
      schemas,
      doc: {
        summary: summary ?? `Use case ${cleanName} in module ${this.moduleName}`,
        description: description ?? `Auto-generated documentation for ${cleanName}`,
        tags: [this.moduleName],
      },
    });

    return this;
  }

  /**
   * Capture Input and Output schemas from decorator metadata or a dummy factory call.
   *
   * Strategy (in order of preference):
   * 1. Class-level: inputClass/outputClass → @Field metadata → build schema.
   * 2. Fallback: call factory({}) and invoke toSchema() on the result (legacy).
   */
  private _extractSchemas<I extends Input, O extends Output>(
    factory: UseCaseFactory<I, O>,
    inputClass?: abstract new (...args: unknown[]) => Input,
    outputClass?: abstract new (...args: unknown[]) => Output,
  ): { input: Record<string, unknown>; output: Record<string, unknown> } {
    let input: Record<string, unknown> = {};
    let output: Record<string, unknown> = {};

    // Strategy 1: class-level via @Field metadata (no instantiation needed)
    if (inputClass) {
      const fields = getFieldMetadata(inputClass);
      if (fields.length > 0) {
        input = new (inputClass as new () => Input)().toSchema();
      }
    }
    if (outputClass) {
      const fields = getFieldMetadata(outputClass);
      if (fields.length > 0) {
        output = new (outputClass as new () => Output)().toSchema();
      }
    }

    // Strategy 2 fallback: dummy factory call
    if (Object.keys(input).length === 0 || Object.keys(output).length === 0) {
      let instance: UseCase<I, O> | undefined;
      try {
        instance = factory({});
      } catch {
        // Factory failed — that's fine.
      }

      if (instance) {
        if (Object.keys(input).length === 0) {
          try {
            input = instance.input.toSchema();
          } catch {
            // toSchema() not available — keep empty.
          }
        }
        if (Object.keys(output).length === 0) {
          try {
            output = instance.output.toSchema();
          } catch {
            // Output not initialised until execute() — expected for most UseCases.
          }
        }
      }
    }

    return { input, output };
  }

  /** @internal — called by ModularApi after the builder callback runs */
  _mount(): void {
    const mountPath = `${this._normalizeBase(this.basePath)}/${this.moduleName}`;
    this.rootRouter.use(mountPath, this.router);
  }

  private _normalizeBase(p: string): string {
    if (!p) return '';
    return p.startsWith('/') ? p : `/${p}`;
  }
}
